File: 12월 4주차/1525 퍼즐.java
import java.util.*;

public class Main {
    static int bfs(String start) {
        String target = "123456789";
        Queue<Pair> queue = new LinkedList<>();
        Set<String> visited = new HashSet<>();

        queue.add(new Pair(start, 0));
        visited.add(start);

        while (!queue.isEmpty()) {
            Pair currentPair = queue.poll();
            String current = currentPair.state;
            int count = currentPair.count;

            if (current.equals(target)) {
                return count;
            }

            int zeroPos = current.indexOf('9');
            int x = zeroPos / 3, y = zeroPos % 3;

            int[] dx = {1, -1, 0, 0};
            int[] dy = {0, 0, 1, -1};

            for (int i = 0; i < 4; i++) {
                int nx = x + dx[i], ny = y + dy[i];
                if (nx >= 0 && nx < 3 && ny >= 0 && ny < 3) {
                    char[] swapped = current.toCharArray();
                    char temp = swapped[zeroPos];
                    swapped[zeroPos] = swapped[nx * 3 + ny];
                    swapped[nx * 3 + ny] = temp;

                    String swappedStr = new String(swapped);
                    if (!visited.contains(swappedStr)) {
                        visited.add(swappedStr);
                        queue.add(new Pair(swappedStr, count + 1));
                    }
                }
            }
        }
        return -1;
    }

    static class Pair {
        String state;
        int count;

        Pair(String s, int c) {
            state = s;
            count = c;
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        StringBuilder board = new StringBuilder();
        for (int i = 0; i < 9; i++) {
            char num = sc.next().charAt(0);
            board.append((num == '0') ? '9' : num);
        }

        int result = bfs(board.toString());
        System.out.println(result);
    }
}

File: 12월 4주차/1525 퍼즐.cpp
#include <iostream>
#include <queue>
#include <unordered_set>
#include <string>
#include <algorithm>
using namespace std;

int bfs(string start) {
    string target = "123456789";
    queue<pair<string, int>> q;
    unordered_set<string> visited;

    q.push(make_pair(start, 0));
    visited.insert(start);

    while (!q.empty()) {
        string current = q.front().first;
        int count = q.front().second;
        q.pop();

        if (current == target) {
            return count;
        }

        int zeroPos = current.find('9');
        int x = zeroPos / 3, y = zeroPos % 3;

        int dx[] = {1, -1, 0, 0}, dy[] = {0, 0, 1, -1};

        for (int i = 0; i < 4; i++) {
            int nx = x + dx[i], ny = y + dy[i];
            if (nx >= 0 && nx < 3 && ny >= 0 && ny < 3) {
                string swapped = current;
                swap(swapped[zeroPos], swapped[nx * 3 + ny]);
                if (visited.find(swapped) == visited.end()) {
                    visited.insert(swapped);
                    q.push(make_pair(swapped, count + 1));
                }
            }
        }
    }
    return -1;
}

int main() {
    string board;
    for (int i = 0; i < 9; i++) {
        char num;
        cin >> num;
        board += (num == '0') ? '9' : num;
    }

    int result = bfs(board);
    cout << result << endl;
    return 0;
}
